// @babylonjs/core/FlowGraph/Blocks/flowGraphBlockNames.js
/**
 * Enum of all block names.
 * Note - if you add a new block, you must add it here, and must add it in the block factory!
 */
export var FlowGraphBlockNames;
(function (FlowGraphBlockNames) {
    FlowGraphBlockNames["PlayAnimation"] = "FlowGraphPlayAnimationBlock";
    FlowGraphBlockNames["StopAnimation"] = "FlowGraphStopAnimationBlock";
    FlowGraphBlockNames["PauseAnimation"] = "FlowGraphPauseAnimationBlock";
    FlowGraphBlockNames["ValueInterpolation"] = "FlowGraphInterpolationBlock";
    FlowGraphBlockNames["SceneReadyEvent"] = "FlowGraphSceneReadyEventBlock";
    FlowGraphBlockNames["SceneTickEvent"] = "FlowGraphSceneTickEventBlock";
    FlowGraphBlockNames["SendCustomEvent"] = "FlowGraphSendCustomEventBlock";
    FlowGraphBlockNames["ReceiveCustomEvent"] = "FlowGraphReceiveCustomEventBlock";
    FlowGraphBlockNames["MeshPickEvent"] = "FlowGraphMeshPickEventBlock";
    FlowGraphBlockNames["PointerEvent"] = "FlowGraphPointerEventBlock";
    FlowGraphBlockNames["PointerDownEvent"] = "FlowGraphPointerDownEventBlock";
    FlowGraphBlockNames["PointerUpEvent"] = "FlowGraphPointerUpEventBlock";
    FlowGraphBlockNames["PointerMoveEvent"] = "FlowGraphPointerMoveEventBlock";
    FlowGraphBlockNames["PointerOverEvent"] = "FlowGraphPointerOverEventBlock";
    FlowGraphBlockNames["PointerOutEvent"] = "FlowGraphPointerOutEventBlock";
    FlowGraphBlockNames["E"] = "FlowGraphEBlock";
    FlowGraphBlockNames["PI"] = "FlowGraphPIBlock";
    FlowGraphBlockNames["Inf"] = "FlowGraphInfBlock";
    FlowGraphBlockNames["NaN"] = "FlowGraphNaNBlock";
    FlowGraphBlockNames["Random"] = "FlowGraphRandomBlock";
    FlowGraphBlockNames["Add"] = "FlowGraphAddBlock";
    FlowGraphBlockNames["Subtract"] = "FlowGraphSubtractBlock";
    FlowGraphBlockNames["Multiply"] = "FlowGraphMultiplyBlock";
    FlowGraphBlockNames["Divide"] = "FlowGraphDivideBlock";
    FlowGraphBlockNames["Abs"] = "FlowGraphAbsBlock";
    FlowGraphBlockNames["Sign"] = "FlowGraphSignBlock";
    FlowGraphBlockNames["Trunc"] = "FlowGraphTruncBlock";
    FlowGraphBlockNames["Floor"] = "FlowGraphFloorBlock";
    FlowGraphBlockNames["Ceil"] = "FlowGraphCeilBlock";
    FlowGraphBlockNames["Round"] = "FlowGraphRoundBlock";
    FlowGraphBlockNames["Fraction"] = "FlowGraphFractBlock";
    FlowGraphBlockNames["Negation"] = "FlowGraphNegationBlock";
    FlowGraphBlockNames["Modulo"] = "FlowGraphModuloBlock";
    FlowGraphBlockNames["Min"] = "FlowGraphMinBlock";
    FlowGraphBlockNames["Max"] = "FlowGraphMaxBlock";
    FlowGraphBlockNames["Clamp"] = "FlowGraphClampBlock";
    FlowGraphBlockNames["Saturate"] = "FlowGraphSaturateBlock";
    FlowGraphBlockNames["MathInterpolation"] = "FlowGraphMathInterpolationBlock";
    FlowGraphBlockNames["Equality"] = "FlowGraphEqualityBlock";
    FlowGraphBlockNames["LessThan"] = "FlowGraphLessThanBlock";
    FlowGraphBlockNames["LessThanOrEqual"] = "FlowGraphLessThanOrEqualBlock";
    FlowGraphBlockNames["GreaterThan"] = "FlowGraphGreaterThanBlock";
    FlowGraphBlockNames["GreaterThanOrEqual"] = "FlowGraphGreaterThanOrEqualBlock";
    FlowGraphBlockNames["IsNaN"] = "FlowGraphIsNaNBlock";
    FlowGraphBlockNames["IsInfinity"] = "FlowGraphIsInfBlock";
    FlowGraphBlockNames["DegToRad"] = "FlowGraphDegToRadBlock";
    FlowGraphBlockNames["RadToDeg"] = "FlowGraphRadToDegBlock";
    FlowGraphBlockNames["Sin"] = "FlowGraphSinBlock";
    FlowGraphBlockNames["Cos"] = "FlowGraphCosBlock";
    FlowGraphBlockNames["Tan"] = "FlowGraphTanBlock";
    FlowGraphBlockNames["Asin"] = "FlowGraphASinBlock";
    FlowGraphBlockNames["Acos"] = "FlowGraphACosBlock";
    FlowGraphBlockNames["Atan"] = "FlowGraphATanBlock";
    FlowGraphBlockNames["Atan2"] = "FlowGraphATan2Block";
    FlowGraphBlockNames["Sinh"] = "FlowGraphSinhBlock";
    FlowGraphBlockNames["Cosh"] = "FlowGraphCoshBlock";
    FlowGraphBlockNames["Tanh"] = "FlowGraphTanhBlock";
    FlowGraphBlockNames["Asinh"] = "FlowGraphASinhBlock";
    FlowGraphBlockNames["Acosh"] = "FlowGraphACoshBlock";
    FlowGraphBlockNames["Atanh"] = "FlowGraphATanhBlock";
    FlowGraphBlockNames["Exponential"] = "FlowGraphExponentialBlock";
    FlowGraphBlockNames["Log"] = "FlowGraphLogBlock";
    FlowGraphBlockNames["Log2"] = "FlowGraphLog2Block";
    FlowGraphBlockNames["Log10"] = "FlowGraphLog10Block";
    FlowGraphBlockNames["SquareRoot"] = "FlowGraphSquareRootBlock";
    FlowGraphBlockNames["CubeRoot"] = "FlowGraphCubeRootBlock";
    FlowGraphBlockNames["Power"] = "FlowGraphPowerBlock";
    FlowGraphBlockNames["Length"] = "FlowGraphLengthBlock";
    FlowGraphBlockNames["Normalize"] = "FlowGraphNormalizeBlock";
    FlowGraphBlockNames["Dot"] = "FlowGraphDotBlock";
    FlowGraphBlockNames["Cross"] = "FlowGraphCrossBlock";
    FlowGraphBlockNames["Rotate2D"] = "FlowGraphRotate2DBlock";
    FlowGraphBlockNames["Rotate3D"] = "FlowGraphRotate3DBlock";
    FlowGraphBlockNames["Transpose"] = "FlowGraphTransposeBlock";
    FlowGraphBlockNames["Determinant"] = "FlowGraphDeterminantBlock";
    FlowGraphBlockNames["InvertMatrix"] = "FlowGraphInvertMatrixBlock";
    FlowGraphBlockNames["MatrixMultiplication"] = "FlowGraphMatrixMultiplicationBlock";
    FlowGraphBlockNames["BitwiseAnd"] = "FlowGraphBitwiseAndBlock";
    FlowGraphBlockNames["BitwiseOr"] = "FlowGraphBitwiseOrBlock";
    FlowGraphBlockNames["BitwiseXor"] = "FlowGraphBitwiseXorBlock";
    FlowGraphBlockNames["BitwiseNot"] = "FlowGraphBitwiseNotBlock";
    FlowGraphBlockNames["BitwiseLeftShift"] = "FlowGraphBitwiseLeftShiftBlock";
    FlowGraphBlockNames["BitwiseRightShift"] = "FlowGraphBitwiseRightShiftBlock";
    FlowGraphBlockNames["LeadingZeros"] = "FlowGraphLeadingZerosBlock";
    FlowGraphBlockNames["TrailingZeros"] = "FlowGraphTrailingZerosBlock";
    FlowGraphBlockNames["OneBitsCounter"] = "FlowGraphOneBitsCounterBlock";
    FlowGraphBlockNames["Branch"] = "FlowGraphBranchBlock";
    FlowGraphBlockNames["SetDelay"] = "FlowGraphSetDelayBlock";
    FlowGraphBlockNames["CancelDelay"] = "FlowGraphCancelDelayBlock";
    FlowGraphBlockNames["CallCounter"] = "FlowGraphCallCounterBlock";
    FlowGraphBlockNames["Debounce"] = "FlowGraphDebounceBlock";
    FlowGraphBlockNames["Throttle"] = "FlowGraphThrottleBlock";
    FlowGraphBlockNames["DoN"] = "FlowGraphDoNBlock";
    FlowGraphBlockNames["FlipFlop"] = "FlowGraphFlipFlopBlock";
    FlowGraphBlockNames["ForLoop"] = "FlowGraphForLoopBlock";
    FlowGraphBlockNames["MultiGate"] = "FlowGraphMultiGateBlock";
    FlowGraphBlockNames["Sequence"] = "FlowGraphSequenceBlock";
    FlowGraphBlockNames["Switch"] = "FlowGraphSwitchBlock";
    FlowGraphBlockNames["WaitAll"] = "FlowGraphWaitAllBlock";
    FlowGraphBlockNames["WhileLoop"] = "FlowGraphWhileLoopBlock";
    FlowGraphBlockNames["ConsoleLog"] = "FlowGraphConsoleLogBlock";
    FlowGraphBlockNames["Conditional"] = "FlowGraphConditionalBlock";
    FlowGraphBlockNames["Constant"] = "FlowGraphConstantBlock";
    FlowGraphBlockNames["TransformCoordinatesSystem"] = "FlowGraphTransformCoordinatesSystemBlock";
    FlowGraphBlockNames["GetAsset"] = "FlowGraphGetAssetBlock";
    FlowGraphBlockNames["GetProperty"] = "FlowGraphGetPropertyBlock";
    FlowGraphBlockNames["SetProperty"] = "FlowGraphSetPropertyBlock";
    FlowGraphBlockNames["GetVariable"] = "FlowGraphGetVariableBlock";
    FlowGraphBlockNames["SetVariable"] = "FlowGraphSetVariableBlock";
    FlowGraphBlockNames["JsonPointerParser"] = "FlowGraphJsonPointerParserBlock";
    FlowGraphBlockNames["CombineVector2"] = "FlowGraphCombineVector2Block";
    FlowGraphBlockNames["CombineVector3"] = "FlowGraphCombineVector3Block";
    FlowGraphBlockNames["CombineVector4"] = "FlowGraphCombineVector4Block";
    FlowGraphBlockNames["CombineMatrix"] = "FlowGraphCombineMatrixBlock";
    FlowGraphBlockNames["CombineMatrix2D"] = "FlowGraphCombineMatrix2DBlock";
    FlowGraphBlockNames["CombineMatrix3D"] = "FlowGraphCombineMatrix3DBlock";
    FlowGraphBlockNames["ExtractVector2"] = "FlowGraphExtractVector2Block";
    FlowGraphBlockNames["ExtractVector3"] = "FlowGraphExtractVector3Block";
    FlowGraphBlockNames["ExtractVector4"] = "FlowGraphExtractVector4Block";
    FlowGraphBlockNames["ExtractMatrix"] = "FlowGraphExtractMatrixBlock";
    FlowGraphBlockNames["ExtractMatrix2D"] = "FlowGraphExtractMatrix2DBlock";
    FlowGraphBlockNames["ExtractMatrix3D"] = "FlowGraphExtractMatrix3DBlock";
    FlowGraphBlockNames["TransformVector"] = "FlowGraphTransformVectorBlock";
    FlowGraphBlockNames["TransformCoordinates"] = "FlowGraphTransformCoordinatesBlock";
    FlowGraphBlockNames["MatrixDecompose"] = "FlowGraphMatrixDecompose";
    FlowGraphBlockNames["MatrixCompose"] = "FlowGraphMatrixCompose";
    FlowGraphBlockNames["BooleanToFloat"] = "FlowGraphBooleanToFloat";
    FlowGraphBlockNames["BooleanToInt"] = "FlowGraphBooleanToInt";
    FlowGraphBlockNames["FloatToBoolean"] = "FlowGraphFloatToBoolean";
    FlowGraphBlockNames["IntToBoolean"] = "FlowGraphIntToBoolean";
    FlowGraphBlockNames["IntToFloat"] = "FlowGraphIntToFloat";
    FlowGraphBlockNames["FloatToInt"] = "FlowGraphFloatToInt";
    FlowGraphBlockNames["Easing"] = "FlowGraphEasingBlock";
    FlowGraphBlockNames["Context"] = "FlowGraphContextBlock";
    FlowGraphBlockNames["ArrayIndex"] = "FlowGraphArrayIndexBlock";
    FlowGraphBlockNames["CodeExecution"] = "FlowGraphCodeExecutionBlock";
    FlowGraphBlockNames["IndexOf"] = "FlowGraphIndexOfBlock";
    FlowGraphBlockNames["FunctionReference"] = "FlowGraphFunctionReference";
    FlowGraphBlockNames["BezierCurveEasing"] = "FlowGraphBezierCurveEasing";
    FlowGraphBlockNames["DataSwitch"] = "FlowGraphDataSwitchBlock";
})(FlowGraphBlockNames || (FlowGraphBlockNames = {}));
