// @babylonjs/core/AudioV2/webAudio/webAudioMainBus.js
import { MainAudioBus } from "../abstractAudio/mainAudioBus.js";
import { _WebAudioBaseSubGraph } from "./subNodes/webAudioBaseSubGraph.js";
/** @internal */
export class _WebAudioMainBus extends MainAudioBus {
    /** @internal */
    constructor(name, engine) {
        super(name, engine);
        this._subGraph = new _WebAudioMainBus._SubGraph(this);
        this.audioContext = engine.audioContext;
    }
    /** @internal */
    async init(options) {
        await this._subGraph.init(options);
        if (this.engine.mainOut) {
            if (!this._connect(this.engine.mainOut)) {
                throw new Error("Connect failed");
            }
        }
        this.engine.addMainBus(this);
    }
    /** @internal */
    dispose() {
        super.dispose();
        this.engine.removeMainBus(this);
    }
    /** @internal */
    get inNode() {
        return this._subGraph.inNode;
    }
    /** @internal */
    get outNode() {
        return this._subGraph.outNode;
    }
    _connect(node) {
        const connected = super._connect(node);
        if (!connected) {
            return false;
        }
        if (node.inNode) {
            this.outNode?.connect(node.inNode);
        }
        return true;
    }
    _disconnect(node) {
        const disconnected = super._disconnect(node);
        if (!disconnected) {
            return false;
        }
        if (node.inNode) {
            this.outNode?.disconnect(node.inNode);
        }
        return true;
    }
    /** @internal */
    getClassName() {
        return "_WebAudioMainBus";
    }
}
_WebAudioMainBus._SubGraph = class extends _WebAudioBaseSubGraph {
    get _downstreamNodes() {
        return this._owner._downstreamNodes ?? null;
    }
};
